Compiler optimizations (take 3, hopefully final)

gkossakowski opened the pull request:
Compiler optimizations prepared by @odersky and me.

All tests for all commits pass:
https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/ (builds 527-553)

This branch has been cooked before in #1157 and #1158. It seems to be ready for serving!

Changes from last iteration:
- fix crasher in first commit (thanks to @paulp)
- fix some typos in a few commits that made build to fail
- dropped @retronym's patch (hopefully I can bring it back today but I'm travelling so no promises)
- squashed two last commits related to making collection methods under 35 bytes

Review by @paulp, @odersky and everyone interested. We need this work to be merged ASAP.

gkossakowski committed:
Don't allocate closure in List.hashCode.

Provided specialized version of hashing algorithm
for List that does not allocate the closure.

This saves us 124k closures allocated in a compiler
when compiling `Vector.scala` from std library.
That's aroudn 5% of all closures allocated for that
compiler run.

odersky committed:
More method hoisting.

Also avoided systematically to map (_.tpe) on parameters in favor of
lazy val paramTypes.

odersky committed:
First half of typed1 reorganization, to make it smaller and faster.

odersky committed:
Second part of typed1 optimization

All cases factored out into separate methods. Cases ordered according to node frequency.

odersky committed:
Method splitting for erasure, genJVM, parsers, scanners, uncurry

Also, added inlined List.foreach

odersky committed:
More method splitting

odersky committed:
Optimizations of isTrivial fields and methods

odersky committed:
Optimization to atOwner

Moved costly hashMap updates into SuperAccessors from general tree transformers and replaced by immutable map.

odersky committed:
No longer uses generic hashCode in UniqueType, since case class hashCode is now the same and faster.

gkossakowski committed:
Make Array creation more efficient in hot places.

Overriden ClassTags in places where we allocate
a lot of Arrays.

odersky committed:
Made hashCode a method on Type to avoid megamorphic dispatch

profiling data shows that accessing the hashCode field has a cost of about 10% of fully hot running times. We speculate that it's megamorphic dispatch which is costing us this much.
Rewrote code so that UniqueType is not an abstract base class with the hashCode field.

Also removed fingerPrinting bloom filter on findMember because it caused complexity and did not gain us anything accdoring to the tests.

Made TypeVar an abstract case class to avoid extractor calls in TypeMap.

odersky committed:
Optimizations to cut down on #closures created

Driven by profile data.

gkossakowski committed:
Mark logging methods in `SymbolTable` with `@inline`.

Those two methods contribute 2% of Function0 allocation
when compiling `Vector.scala`.

gkossakowski committed:
Rework synchronization scheme of `UndoLog` for performance.

Introduced explicit locking mechanism that allows us to
make public methods of `UndoLog` that take thunks as
arguments final and thus `@inline`. Once inliner is able
to cope with inlining methods with exception handlers
we'll be able to get rid of significant (around 5%)
number of thunk allocation.

odersky committed:
Moved five methods from LinearSeqOptimized to List

Curiously, profiling showed that these methods (in particular map) got slower when specialized in LinearSeqOptimized from their generic impl in TraversableLike.
I speculate that the virtual calls to head/tail/nonEmpty are more expensive than the closure creation which we save (Note that foreach IS specialized in List,
so head/tail are monomorphic, even though isEmpty still has two implementations.

If we confirm a speedup in this commit it would make sense to specialize more methods in LinearSeqOptimized again in List.

odersky committed:
Specialize hashCode in TypeRef from the generic MurmurHashCode3 implementation.

odersky committed:
Optimizations to SymTree and transform

Two optimizations:

(1) Make SymTree a class rather than a trait to allow faster access to symbol
(2) Split itransform into several smaller methods and order cases according to expected frequency.

gkossakowski committed:
Manaul inling of `UndoLog.undoUnless`.

Inline `UndoLog.undoUnless` into `isSubType`
and `isSameType` methods. They are responsible
for 20% of all Function0 allocations.

I had to do manual inlining because inlier
doesn't cope with exception handlers well.

Let it be noted that I almost cried while doing
this ugly change. I'll want to see inliner
being fixed so we can revert this change ASAP.

odersky committed:
Reverted addition of 5 specialized method from LinearSeqOptimized to List.

It seems any gains were more then compensated by the loss of making map polymorphic, probably preventing inlining decisions in Hotspot.

odersky committed:
Partial revert of itransform commit.

Profile data actually showed a small slowdown in steady state. We keep the reordering but merge all submethods into one.

odersky committed:
Made all statistic code disappear unless built with Statistics.canEnable = true

odersky committed:
Some more closure hositing/elimination

odersky committed:
Reverted closure hoisting except for functions returning a Boolean result

If our theory wrt map is correct (i.e. that it gets inlined by JVM), then closure hoisting is counter-productive because it migh well prevent the closure from being inlined. This commit removes all hoisted closures that are passed to map and leaves only those boolean closures that are passed to exists and forall.

I should have split the early optimizations including closures into separate commits. As that train has left, I am now reverting some bits to see whether the reverts affect performance at all, and in what direction.

odersky committed:
Slightly less drastic disabling of statistics.

All statistics code is now subject to a switch at each use site, as opposed to a compile-time constant before. We should veryfy that this does not affect performance. Statistics is a useful ability to have for trouble-shooting compile time speeds on specific code bases. But we should not pay a lot for having it on normal runs.

odersky committed:
Replaced isTrivial lazy vals by custom scheme to save space.

odersky committed:
Splitting large methods into smaller ones.

odersky committed:
Compilespeed improvements: Exists arguments and others

It turns out that exists is not inlinable, even if put into List.
We try to eliminate or hoist most closures passed to exists in Types.
There are some other small improvements as well.

--

(@gkossakowski): This commit contains also a fix to crasher
prepared by @paulp. I squashed that commit and kept the
test-case that came with it.

retronym commented:
Please leave out my patch. I'll submit it as a separate request.

retronym commented:
Do you have numbers from your performance test rig to show how each commit affects the performance?

gkossakowski commented:
Not yet. I'm waiting for rebased commits to finish building (i did small corrections in commit messages) and then I'll run benchmarks. They will take a couple of hours (around 10) to finish for that number of commits.

However, our testing with https://github.com/pk11/playbench (11k LoC extracted from Play) showed that we are slightly faster than 2.9.2. That's a huge performance win because 2.10.x branch was significantly slower than 2.9.2 before we jumped on performance work.

PS. I intent t make my tooling public but it's not there yet unfortunately. I'm afraid it will take me 1-2 months before it can go public.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/245/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/952/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/245/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/952/

gkossakowski committed:
Attempts to improve inlining behavior for map, flatMap.

Reduced method size of map/flatMap to under 35 bytes to make them
inlinable. Also specialized filterNot code so that it does not create
a second closure. Finally, avoided default argument for sizeHint to
reduce size of methods calling it.

gkossakowski commented:
@hubertp: Fixed typo. Thanks!

gkossakowski commented:
@retronym: Sure. I'm happy to review your pull request. BTW. If you have some spare time I'd be grateful if you could have a look into this one. I feel it's fairly safe but the more eye balls the better.

jsuereth commented:
Ok, we're passing.  Should I wait for a full LGTM, or can I merge?  I need to cut a milestone before my COB today, or it ain't happening for a bit while I travel the atlantic.

gkossakowski commented:
Well, most of work has been done by @odersky so I'll review his work now. I did partial reviews as we were progressing. My work is fairly trivial apart from b2f8acd and 82c6168 that I'd appreciate if you could review it.

retronym commented on src/reflect/scala/tools/nsc/io/ZipArchive.scala:
```diff
@@ -96,14 +96,25 @@ abstract class ZipArchive(override val file: JFile) extends AbstractFile with Eq
+    //   parent.entries(baseName(path)) = dir
+    //   dir
+    // })
+    dirs get path match {
```
`scala.Map` is pretty allocation heavy with `Tuple2` and `Some`. 

If profiling still shows a hotspot , you could switch to `java.util.Map`, or something like:

``` scala
final class FastHashMap[A, B >: Null] extends scala.collection.mutable.HashMap[A, B] {
  override def contains(key: A) = findEntry(key) != null

  def getOrNull(key: A): B = {
    val entry = findEntry(key)
    if (entry == null) null else entry.value
  }
}
```

gkossakowski replied:
We focused on closures because when we started there were around 2.2m of `Function0` and `Function1` allocated when we compiled `Vector.scala` and only "a few" hundred thousand of Tuple2. Yeah, in the second pass we should optimize collection operations.

I'll treat such comments as notes for future work. I might be able to get to work on that next week. I think this is not a blocker for pull request to be accepted, right?

retronym replied:
Yep, it's just a pointer for the next round.

retronym replied:
Looks like someone beat me to it: b924c41.

gkossakowski commented:
Ok. I went through all commits and I haven't seen anything horribly broken. Even if we merge this pull request I'll be coming back to changes, benchmarking them and refining them.

I validated that all commits build and pass all tests so I think it's good enough to give LGTM from me. Let's merge it before we run into merge conflicts.
